<a href="https://colab.research.google.com/github/Mohit03singh/Customer_segmentation_using_unsupervised_learning/blob/main/Customer_segmentation_using_unsupervised_learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**NAME** : Mohit

**COHORT** : SANTIAGO

**TYPE** : INDIVIDUAL

GITHUB LINK :

### **What is customer segmentation?**

#### Customer segmentation is a way to split customers into groups based on certain characteristics that those customers share. All customers share the common need of your product or service, but beyond that, there are distinct demographic differences (i.e., age, gender) and they tend to have additional socio-economic, lifestyle, or other behavioral differences that can be useful to the organization.


### **Why segment customers?**

#### Segmentation allows marketers to better tailor their marketing efforts to various audience subsets. Those efforts can relate to both communications and product development. Specifically, segmentation helps a company:

  * Create and communicate targeted marketing messages that will resonate with specific groups of customers, but not with others (who will receive messages tailored to their needs and interests, instead).
  * Select the best communication channel for the segment, which might be email, social media posts, radio advertising, or another approach, depending on the segment.
  * Identify ways to improve products or new product or service opportunities.
  * Establish better customer relationships.
  * Test pricing options.
  * Focus on the most profitable customers.
  * Improve customer service.
  * Upsell and cross-sell other products and services.

### <b> Data Description: </b>

#### <b>Attribute Information: </b>

* #### InvoiceNo: Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation.
* #### StockCode: Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each distinct product.
* #### Description: Product (item) name. Nominal.
* #### Quantity: The quantities of each product (item) per transaction. Numeric.
* #### InvoiceDate: Invice Date and time. Numeric, the day and time when each transaction was generated.
* #### UnitPrice: Unit price. Numeric, Product price per unit in sterling.
* #### CustomerID: Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer.
* #### Country: Country name. Nominal, the name of the country where each customer reside

In [1]:
#IMPORT THE LIBRARIES

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from datetime import datetime
import datetime as dt

from sklearn import preprocessing
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import silhouette_score
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_samples, silhouette_score
from sklearn import metrics

In [2]:
# load the dataset

df = pd.read_csv('/content/Online_Retail_-_Copy[1].csv')

In [3]:
cust_data = df.copy()

In [4]:
cust_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

    InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/10 8:26       2.55     17850.0  United Kingdom  
1  12/1/10 8:26       3.39     17850.0  United Kingdom  
2  12/1/10 8:26       2.75     17850.0  United Kingdom  
3  12/1/10 8:26       3.39     17850.0  United Kingdom  
4  12/1/10 8:26       3.39     17850.0  United Kingdom

In [5]:
cust_data.tail()

InvoiceNo StockCode                      Description  Quantity  \
541904    581587     22613      PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899     CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254    CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255  CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138    BAKING SET 9 PIECE RETROSPOT          3   

          InvoiceDate  UnitPrice  CustomerID Country  
541904  12/9/11 12:50       0.85     12680.0  France  
541905  12/9/11 12:50       2.10     12680.0  France  
541906  12/9/11 12:50       4.15     12680.0  France  
541907  12/9/11 12:50       4.15     12680.0  France  
541908  12/9/11 12:50       4.95     12680.0  France

In [6]:
cust_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [7]:
cust_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [8]:
cust_data.shape

(541909, 8)

In [9]:
cust_data.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [10]:
cust_data.dtypes

InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float64
Country         object
dtype: object

We can see that invoice no is in object format we need to convert that to date format

# **Handling missing values and Duplicates in the dataset**

In [11]:
#handling duplicate values
cust_data.duplicated().sum()

5268

We can see that there are 5268 duplicate values we need to eliminate them

In [12]:
#handling null values
cust_data.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

We can see that there are 1454 and 135080 null values in Description and cutomer id feature we need to eliminate them

Removal of duplicates and null values

In [13]:
cust_data.drop_duplicates(inplace = True)
cust_data.dropna(inplace = True)

In [14]:
cust_data.duplicated().sum()

0

In [15]:
cust_data.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [16]:
cust_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 401604 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    401604 non-null  object 
 1   StockCode    401604 non-null  object 
 2   Description  401604 non-null  object 
 3   Quantity     401604 non-null  int64  
 4   InvoiceDate  401604 non-null  object 
 5   UnitPrice    401604 non-null  float64
 6   CustomerID   401604 non-null  float64
 7   Country      401604 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 27.6+ MB


In [17]:
cust_data.shape

(401604, 8)

Earlier the shape of the dataset was (541909, 8) and now after removal of null and duplicate values it has been reduced to (401604,8)

# **Removal of Cancelled Orders**

In [18]:
cust_data['InvoiceNo'] = cust_data['InvoiceNo'].astype('str')
cust_data[cust_data['InvoiceNo'].str.startswith('C')]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

          InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/10 9:41      27.50     14527.0  United Kingdom  
154      12/1/10 9:49       4.65     15311.0  United Kingdom  
235     12/1/10 10:24       1.65     17548.0  United Kingdom  
236     12/1/10 10:24       0.29     17548.0  United Kingdom  
237     12/1/10 10:24       0.29     17548.0  United Kingdom  
...               ...        ...         ...             ...  
540449   12/9/11 9:57       0.83     14397.0  United Kingdom  
541541  12/9/11 10:28     224.69     15498.0  United Kingdom  
541715  12/9/11 11:57      10.95     15311.0  United Kingdom  
541716  12/9/11 11:58       1.25     17315.0  United Kingdom  
541717  12/9/11 11:58       1.25     17315.0  United Kingdom  

[8872 rows x 8 columns]

We can see there are 8872 orders which are cancelled and still present in the dataset hence we need to remove them

In [19]:
cust_data = cust_data[~cust_data['InvoiceNo'].str.startswith('C')]


In [20]:
cust_data.shape

(392732, 8)

Earlier the shape of the dataset was (401604,8) and now after removal of cancelled orders it has been reduced to (392732, 8)

In [21]:
cust_data.describe()

Quantity      UnitPrice     CustomerID
count  392732.000000  392732.000000  392732.000000
mean       13.153718       3.125596   15287.734822
std       181.588420      22.240725    1713.567773
min         1.000000       0.000000   12346.000000
25%         2.000000       1.250000   13955.000000
50%         6.000000       1.950000   15150.000000
75%        12.000000       3.750000   16791.000000
max     80995.000000    8142.750000   18287.000000

We can see that the min values are now positive but the unitprice here cannot be 0 because it results in free distribution

In [22]:

cust_data[cust_data['UnitPrice']==0]

InvoiceNo StockCode                          Description  Quantity  \
9302      537197     22841         ROUND CAKE TIN VINTAGE GREEN         1   
33576     539263     22580         ADVENT CALENDAR GINGHAM SACK         4   
40089     539722     22423             REGENCY CAKESTAND 3 TIER        10   
47068     540372     22090              PAPER BUNTING RETROSPOT        24   
47070     540372     22553               PLASTERS IN TIN SKULLS        24   
56674     541109     22168        ORGANISER WOOD ANTIQUE WHITE          1   
86789     543599    84535B         FAIRY CAKES NOTEBOOK A6 SIZE        16   
130188    547417     22062  CERAMIC BOWL WITH LOVE HEART DESIGN        36   
139453    548318     22055   MINI CAKE STAND  HANGING STRAWBERY         5   
145208    548871     22162          HEART GARLAND RUSTIC PADDED         2   
157042    550188     22636   CHILDS BREAKFAST SET CIRCUS PARADE         1   
187613    553000     47566                        PARTY BUNTING         4   
198383    554037     22619            SET OF 6 SOLDIER SKITTLES        80   
279324    561284     22167           OVAL WALL MIRROR DIAMANTE          1   
282912    561669     22960             JAM MAKING SET WITH JARS        11   
285657    561916         M                               Manual         1   
298054    562973     23157           SET OF 6 NATIVITY MAGNETS        240   
314745    564651     23270     SET OF 2 CERAMIC PAINTED HEARTS         96   
314746    564651     23268  SET OF 2 CERAMIC CHRISTMAS REINDEER       192   
314747    564651     22955             36 FOIL STAR CAKE CASES        144   
314748    564651     21786                   POLKADOT RAIN HAT        144   
358655    568158      PADS           PADS TO MATCH ALL CUSHIONS         1   
361825    568384         M                               Manual         1   
379913    569716     22778                   GLASS CLOCHE SMALL         2   
395529    571035         M                               Manual         1   
420404    572893     21208          PASTEL COLOUR HONEYCOMB FAN         5   
436428    574138     23234        BISCUIT TIN VINTAGE CHRISTMAS       216   
436597    574175     22065       CHRISTMAS PUDDING TRINKET POT         12   
436961    574252         M                               Manual         1   
439361    574469     22385            JUMBO BAG SPACEBOY DESIGN        12   
446125    574879     22625                   RED KITCHEN SCALES         2   
446793    574920     22899         CHILDREN'S APRON DOLLY GIRL          1   
446794    574920     23480       MINI LIGHTS WOODLAND MUSHROOMS         1   
454463    575579     22437        SET OF 9 BLACK SKULL BALLOONS        20   
454464    575579     22089        PAPER BUNTING VINTAGE PAISLEY        24   
479079    577129     22464          HANGING METAL HEART LANTERN         4   
479546    577168         M                               Manual         1   
480649    577314     23407       SET OF 2 TRAYS HOME SWEET HOME         2   
485985    577696         M                               Manual         1   
502122    578841     84826       ASSTD DESIGN 3D PAPER STICKERS     12540   

           InvoiceDate  UnitPrice  CustomerID         Country  
9302     12/5/10 14:02        0.0     12647.0         Germany  
33576   12/16/10 14:36        0.0     16560.0  United Kingdom  
40089   12/21/10 13:45        0.0     14911.0            EIRE  
47068     1/6/11 16:41        0.0     13081.0  United Kingdom  
47070     1/6/11 16:41        0.0     13081.0  United Kingdom  
56674    1/13/11 15:10        0.0     15107.0  United Kingdom  
86789    2/10/11 13:08        0.0     17560.0  United Kingdom  
130188   3/23/11 10:25        0.0     13239.0  United Kingdom  
139453   3/30/11 12:45        0.0     13113.0  United Kingdom  
145208    4/4/11 14:42        0.0     14410.0  United Kingdom  
157042   4/14/11 18:57        0.0     12457.0     Switzerland  
187613   5/12/11 15:21        0.0     17667.0  United Kingdom  
198383   5/20/11 

There are some values which has 0 unit price we need to drop them.

In [23]:
cust_data = cust_data[cust_data['UnitPrice']>0]

cust_data.describe()

Quantity      UnitPrice     CustomerID
count  392692.000000  392692.000000  392692.000000
mean       13.119702       3.125914   15287.843865
std       180.492832      22.241836    1713.539549
min         1.000000       0.001000   12346.000000
25%         2.000000       1.250000   13955.000000
50%         6.000000       1.950000   15150.000000
75%        12.000000       3.750000   16791.000000
max     80995.000000    8142.750000   18287.000000

In [24]:
cust_data['UnitPrice'].describe()

count    392692.000000
mean          3.125914
std          22.241836
min           0.001000
25%           1.250000
50%           1.950000
75%           3.750000
max        8142.750000
Name: UnitPrice, dtype: float64

We can see that the min value of unit price is slightly changed..

cust_data.shape

The data is almost cleaned and ready for EDA

# **Feature Engineering**

#### Feature engineering is a machine learning technique that leverages data to create new variables that aren’t in the training set. It can produce new features for both supervised and unsupervised learning, with the goal of simplifying and speeding up data transformations while also enhancing model accuracy.

In [ ]:
cust_data['InvoiceDate'] = cust_data['InvoiceDate'].apply(lambda x: pd.to_datetime(x))


In [ ]:
cust_data['InvoiceDate'].dtypes

In [ ]:
# create some new features from invoice date
cust_data['day'] = cust_data['InvoiceDate'].dt.day_name()
cust_data['year'] = cust_data['InvoiceDate'].apply(lambda x: x.year)
cust_data['month_num'] = cust_data['InvoiceDate'].apply(lambda x: x.month)
cust_data['day_num'] = cust_data['InvoiceDate'].apply(lambda x: x.day)
cust_data['hour'] = cust_data['InvoiceDate'].apply(lambda x: x.hour)
cust_data['minute'] = cust_data['InvoiceDate'].apply(lambda x: x.minute)
cust_data['month'] = cust_data['InvoiceDate'].dt.month_name()

In [ ]:
cust_data['Total_amount'] = cust_data['UnitPrice']*cust_data['Quantity']

# **Exploratory Data Analysis**


What is Exploratory Data Analysis (EDA) ?

EDA is a phenomenon under data analysis used for gaining a better understanding of data aspects like: – main features of data – variables and relationships that hold between them – identifying which variables are important for our problem We shall look at various exploratory data analysis methods like:

Descriptive Statistics, which is a way of giving a brief overview of the dataset we are dealing with, including some measures and features of the sample Grouping data [Basic grouping with group by]

ANOVA, Analysis Of Variance, which is a computational method to divide variations in an observations set into different components.

Correlation and correlation methods

# **1. MOST SOLD PRODUCTS**

In [ ]:
most_sold = cust_data['Description'].value_counts().reset_index()
most_sold.rename(columns = {'index':'Product_Name','Description':'No_of_items_Count'}, inplace =True)
most = most_sold.head()

In [ ]:
most

In [ ]:
sns.barplot(x='Product_Name',y='No_of_items_Count', data = most,palette='viridis')
plt.title("Most sold products")
plt.xticks(rotation = 90)
plt.show()

The Five most sold products are:

1.WHITE HANGING HEART T-LIGHT HOLDER(2016)

2.REGENCY CAKESTAND 3 TIER(1713)

3.JUMBO BAG RED RETROSPOT(1615)

4.ASSORTED COLOUR BIRD ORNAMENT(1395)

5.PARTY BUNTING(1389)

In [ ]:
least = most_sold.tail()

In [ ]:
sns.barplot(x='Product_Name',y='No_of_items_Count', data = least,palette='viridis')
plt.title("Least sold products")
plt.xticks(rotation = 90)
plt.show()